File: Java/TreeNode.java
import java.util.ArrayList;
import java.util.function.Consumer;


public class TreeNode<T> {
    private T value;
    private ArrayList<TreeNode<T>> children = new ArrayList<TreeNode<T>>();
    
    TreeNode(T v)
    {
        value = v;
    }
    
    void add (TreeNode<T> child)
    {
        children.add(child);
    }
    
    T getValue()
    {
        return value;
    }
    
	void depthFirst(Consumer<TreeNode<T>> visit)
	{
		visit.accept(this);
		for (TreeNode<T> node : children)
			node.depthFirst(visit);
	}
	
	void breadthFirst(Consumer<TreeNode<T>> visit)
	{
		ArrayList<TreeNode<T>> queue = new ArrayList<TreeNode<T>>();
		
		visit.accept(this);
		for (TreeNode<T> child : children)
			queue.add(child);
		
		int idx = 0;
		while (idx < queue.size())
		{
			TreeNode<T> node = queue.get(idx);
			visit.accept(node);
			idx += 1;
			for (TreeNode<T> grandchild : node.children)
				queue.add(grandchild);	
		}
		
	}
    
	public void printByLevel()
	{
		ArrayList<TreeNode<T>> queue = new ArrayList<TreeNode<T>>();
        
		queue.add(this);
		int nextLevel = 1;
        while (queue.size() != 0)
        {
            nextLevel = printCurrentLevel(nextLevel, queue);
            System.out.println();
        }       
	}
    
    private int printCurrentLevel(int toBePrinted, ArrayList<TreeNode<T>> queue)
    {
        int nextLevel = 0;
        for (int i=0; i<toBePrinted; i+=1)
        {
            TreeNode<T> node = queue.remove(0);
            System.out.print(node.getValue() + " ");
            for (TreeNode<T> child : node.children)
            {
                nextLevel += 1;
                queue.add(child);
            }
        }
        return nextLevel;
    }
    
    
}

File: Swift/Trees/Trees/TreeNode.swift
//
//  TreeNode.swift
//  Trees
//

import Foundation

public class TreeNode<T> {
  public var value: T
  public var children: [TreeNode] = []

  public init(_ value: T) {
    self.value = value
  }
  
  public func add(_ child: TreeNode) {
    children.append(child)
  }
}

// #####     #####      #####     #####     #####
//      #####      #####     #####     #####
// #####     #####      #####     #####     #####

extension TreeNode {
  public func forEachDepthFirst(visit: (TreeNode) -> Void) {
    visit(self)
    children.forEach {
      $0.forEachDepthFirst(visit: visit)
    }
  }
}

// #####     #####      #####     #####     #####
//      #####      #####     #####     #####
// #####     #####      #####     #####     #####

extension TreeNode {
  public func forEachBreadthFirst(visit: (TreeNode) -> Void)
  {
    var queue : [TreeNode] = []
    
    visit(self)
    for child in children {
      queue.append(child)
    }
    
    var idx = 0
    while (idx < queue.count)
    {
      let node = queue[idx]
      visit(node)
      idx += 1
      for grandChild in node.children
      {
        queue.append(grandChild)
      }
    }
  }
}
